### ISSUE TITLE
Suggestion on adding an ExecReduce() method

### ISSUE THREAD

**Issue by gty929** [bitfield/script#86]
Hi, I'm proposing to add an `ExecReduce()` method, which should be similar to Array.prototype.reduce() in JavaScript. I come up with this idea after reading issue #37 . The method could be a general way to solve this type of issue, as well as many other real-world problems. Also, the method fits in the current program interfaces, and does not increase the learning cost by much. It can cover some of the shortages in the current `EachLine()` and `ExecForEach()` methods, as will be elaborated below based on several use cases.

The interface of the `ExecReduce()` method resembles `ExecForEach()`: 

`func (p *Pipe) ExecReduce(cmdTpl string, initVals ...string) *Pipe`

And here's a design on how it should work:
ExecReduce runs the supplied "reducer" command on each line of input, passing in the return value from the calculation on the preceding line. The final result of running the reducer across lines is a single value.
The command string is interpreted as a Go template, which parses {{.First}} as the previously reduced value and {{.Second}} as the current line value. By default, the reduction starts from the first two lines, whereas a user may also specify one or more strings as the initial values. The function does not jump over empty lines.
The first command that results in an error will set the pipe's error status accordingly, and no subsequent commands will be run.

For instance, the use case of Issue #37 can be implemented as `ExecReduce("echo '{{.First}}, {{.Second}}'")`

Here's another simple use case: Given a file with one number per line, sum up all the numbers. With `ExecReduce()`, we just need to define how to add two numbers, and the function will add up all the numbers iteratively.
A sample implementation would be `ExecReduce("bash -c 'echo $(({{.First}} + {{.Second}}))'")`

A more concrete use case could be: I have an account with an initial balance of $1000, and a file where each line defines an operation. For example,

`Transfer in $100;`
`Add 1% interest;`
`Transfer out $80;`

The goal is to calculate the final balance after all the operations are performed sequentially.

Suppose that I've written a script `performOp.sh` which, taking in the current balance and a single operation string, echos the balance after the operation. Then, I can get the final balance by simply running `ExecReduce("./performOp.sh {{.First}} {{.Second}}", "1000")`

Finally, I want to compare `ExecReduce()` with the existing methods `EachLine()` and `ExecForEach()`.
Again, let's look at a concrete problem: given a file with many lines, print out the line that has the maximum word count. (In reality, 'word count' can be changed into other functions -- it's just for simplicity here)
With `ExecReduce()`, the implementation could be:
`ExecReduce("bash -c '[[ $(echo {{.First}} | wc -w) -ge $(echo {{.Second}} | wc -w) ]] && echo {{.First}} || echo {{.Second}}'")`

The problem could also be tackled with the `EachLine()` method, for example:
```
EachLine(func(line string, out *strings.Builder) {
	if len(strings.Split(out.String(), " ")) < len(strings.Split(line, " ")) {
		out.Reset()
		out.WriteString(line + "\n")
	}
})
```
The main drawback, however, is that we need to reimplement bash commands. While `wc -w` seems simple to implement, one might still make mistakes. The codes above, for instance, would be wrong if `\t` is used instead of common white spaces. (The correct implementation should use `strings.Fields`.) There are also for sure many use cases that require more complicated commands. Besides, the string builder doesn't seem quite elegant here.

In theory, `ExecForEach()` can solve the problem as well. One may calculate the word count for each line, execute another command that finds the index of the line with the max count, and finally query that line based on the index. The implementation would be quite tedious, though, and probably needs more than one pipeline.

In general, the `ExecReduce()` method seems great for problems that require aggregation and chaining of the lines, which the existing methods do not support well. I'm eager to receive any kind of advice. If approved, I can make a pull request soon, including the implementation, tests, examples, and documentation.

**Comment by gty929** [comment]
`func (p *Pipe) ExecReduce(cmdTpl string, initVals ...string) *Pipe` seems to be a better interface.

**Comment by bitfield** [comment]
Hey @gty929, thanks for the suggestion! Much appreciated.

It would be interesting to evaluate this design against some real-world use cases—I'm not saying your "operations on a bank account" isn't realistic, but it doesn't seem on the face of it the kind of thing that someone would want to do with a shell script. Maybe we can come up with some suggestions for such programs, and see how they would work both with and without the suggested `ExecReduce` method.

I think my general approach would be to avoid shelling out to Unix commands to do _computations_: that's the sort of thing Go is good at, so a Go function would be the appropriate way to do that in a Go program. So I'm having trouble envisaging a program using `ExecReduce` to shell out to Unix commands to compute something that wouldn't be better computed using a Go function supplied to `EachLine`, or similar. Maybe you can help?

**Comment by gty929** [comment]
@bitfield Thank you for your comment. Let's say I have a list of IP addresses for a website, and I just need to find out one that works. With `ExecReduce` the code will look like,
```
ExecReduce("bash -c '[[ '{{.First}}' = 'None' ]] && ping -c 4 -W 1 {{.Second}} >> ping.log && echo {{.Second}} || echo {{.First}}'", "None")
```
The reducer will stop pinging once it finds an available address.
The existing methods cannot handle this problem well. With `EachLine`, I cannot use the `ping` command. With `ExecForEach`, I have to ping all the addresses and still need to think of a way to aggregate the results. With `Exec` or a traditional shell script, I cannot read the IP addresses directly from a previous pipe function.

**Comment by gty929** [comment]
Also, do you think we can add a concurrent version of `EachLine()` and `ExecForEach`, which allows functions to be executed in parallel? I guess this is what Go is best at. For example, I can save time by pinging multiple addresses at the same time. I may want to test if my webserver can handle concurrent requests correctly.

**Comment by bitfield** [comment]
It's an interesting idea, but I don't find the IP address example particularly easy to understand, and I don't think others would either. I've never seen a shell script that works this way.

Indeed, I wouldn't use `ping` for this: I'd probably use a Go library like [`fastping`](https://github.com/tatsushid/go-fastping). It supports pinging multiple addresses at once and triggering a callback for whichever one answers.

In general, I think it's fair to say that the intended aim of `script` is not to make it _easier_ to shell out to Unix commands, but to make it largely _unnecessary_. Most pipe methods are replacements of traditional commands like `sed` and `grep`. So if we wanted to write `script` programs that ping IP addresses, we would probably write a `Ping` method on the pipe instead.

I'd be interested to hear your suggestions about adding concurrency support to `script`: do open another issue to track that, since it's a topic all of its own. What kind of concurrent operations would make sense? What would the API look like? What are some typical programs that we'd like to write using the putative concurrency features?

**Comment by bitfield** [comment]
Closing as per comments.
